Source/WebCore/dom/ShadowRoot.cpp

```diff
@@ -125,8 +125,10 @@ Node::InsertedIntoAncestorResult ShadowRoot::insertedIntoAncestor(InsertionType
 void ShadowRoot::removedFromAncestor(RemovalType removalType, ContainerNode& oldParentOfRemovedTree)
 {
     DocumentFragment::removedFromAncestor(removalType, oldParentOfRemovedTree);
-    if (removalType.disconnectedFromDocument)
+    if (removalType.disconnectedFromDocument) {
+        // Unable to ref the document as it may have started destruction.
```

Comment:
How does the static analyzer deal with this? It keeps coming up so I am wondering what our systematic way of dealing with the safety of such code should be.

Reply:
The static analyzer had way too many issues to be truly helpful the last time I used it. My understanding is that Ryosuke is trying to improve it.

Ryosuke's plan for these particular cases was to allow ref'ing objects that started running their destructor:
- https://github.com/WebKit/WebKit/pull/8748

However, the patch was rejected and I do know that we have an alternative plan yet. I plan on eventually looking at refactoring the code to see if some of this code can be stopped from running when the document has started destruction, which is why I am annotating them with a comment.

I also heard that we may get support for pragmas to have the analyzer ignore part of the code in the future (but likely not soon).